### DISCUSSION TITLE
Possibility to add contenttype versioning

### DISCUSSION THREAD

**Discussion by gentledepp** [OrchardCMS/OrchardCore#15874] [Q&A]
Hi!
I wonder if it was possible to introduce some kind of contenttype versioning.

Scenario:
Version 1) We define a contenttype "issue" which contains fields for a user that this issue is assigned to, some description and also some date, when this issue is to be resolved.
Version 2) we create a new version - version 2 which has another content part that requires the users to add  some taxonomies to the content item (required field)

=> now if some content item, which is still based on version 1 is opened, the taxonomies field should not show up, as it is from the v2 definition. However, it should be possible for a user to manually (explicitly) upgrade to version 2, thereby, having to fill in the new mandatory taxonomy field.

Version 3) we decide that the description is not good enough, as it is only text, so we remove it and add a contentpart that adds some markdown editor.

Now when we create a new issue, we want to see that markdown editor of course.
But when opening an old content item, that was created with the v1 or v2 content type, the old description text should still be visible (and not the markdown editor).

I hope you see where I want to go with this.
So as I know, that this is not currently implemented, would it be possible to do so?
(especially not removing any data from a content item, just because it was removed from the content type)

Also, all content type definitions are currently stored in one single document. I sincerely doubt that this can stay as is when it comes to versioning. I just won't scale.

So please help me figure this out! Thanks for your insights!

**Comment by sebastienros** [comment]
This could be handled with a custom type name constructed like "ContentType:Version" which would not break any current API. And then each content item is still linked to a specific version. However we would need to handle this format in many places, editors, driver coordinators ... And maybe expose the version as a property on ContentTypeDefinitionRecord. 

Could be complicated also if you want to handle a single template for any version, you'd have lots of conditions. Or create an alternate using the version number.

This will surely add some complexity in any code that deals with content types. You can try to do something if you have some time to spare, I can't promise we would take it as it's not a features that has been requested so far, and it would add complexity.

**Comment by gentledepp** [comment]
Thank you for your insights.
I truly understand, that this is not an "every day" scenario for what orchard 2 was/is built.
So don't get me wrong: I do not expect you guys to squeeze in every feature, that I think would fit.

I only want to make sure, that the extension points exist, so that guys like me can actually use orchard.

**Comment by agriffard** [comment]
Related to #506
